### PR TITLE
AHRS - use compass singleton

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp [hunk 1]
```diff
@@ -440,14 +440,15 @@ AP_AHRS_DCM::drift_correction_yaw(void)
         /*
           we are using compass for yaw
          */
-        if (_compass->last_update_usec() != _compass_last_update) {
-            yaw_deltat = (_compass->last_update_usec() - _compass_last_update) * 1.0e-6f;
-            _compass_last_update = _compass->last_update_usec();
+        Compass &_compass = AP::compass();
```
Comment (khancyr) [inline libraries/AP_AHRS/AP_AHRS_DCM.cpp]: put this sooner ?

Comment (peterbarker) [reply libraries/AP_AHRS/AP_AHRS_DCM.cpp]: It's scoped as tightly as is sensible IMO.... it's not needed my most of the code in the function.

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp [hunk 2]
```diff
@@ -235,20 +235,21 @@ void NavEKF3_core::writeOptFlowMeas(uint8_t &rawFlowQuality, Vector2f &rawFlowRa
 // check for new magnetometer data and update store measurements if available
 void NavEKF3_core::readMagData()
 {
-    if (!_ahrs->get_compass()) {
+    const Compass &compass = AP::compass();
```
Comment (OXINARF) [inline libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp]: Nitpicking, but can we make EKF2 and EKF3 be consistent? Looks like EKF2 code is after the comment below, EKF3 code is before the comment - I prefer the EKF3 way, but I just care that changes are consistent.

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp (hunk 1 above)
Comment (OXINARF) [inline libraries/AP_AHRS/AP_AHRS_DCM.cpp]: Can we remove the underscore prefix? You already have to change the lines below so...

File: APMrover2/APMrover2.cpp [hunk 3]
```diff
@@ -193,7 +193,7 @@ void Rover::gcs_failsafe_check(void)
 void Rover::update_compass(void)
 {
     if (g.compass_enabled && compass.read()) {
-        ahrs.set_compass(&compass);
+        ahrs.update_orientation();
```
Comment (OXINARF) [inline APMrover2/APMrover2.cpp]: This is new behaviour, what's the reason behind it? (applies to all vehicles and Replay)

File: APMrover2/sensors.cpp [hunk 4]
```diff
@@ -13,7 +13,7 @@ void Rover::init_compass()
         hal.console->printf("Compass initialisation failed!\n");
         g.compass_enabled = false;
     } else {
-        ahrs.set_compass(&compass);
+        ahrs.update_orientation();
```
Comment (OXINARF) [inline APMrover2/sensors.cpp]: Same.

File: libraries/DataFlash/LogFile.cpp [hunk 5]
```diff
@@ -1476,19 +1478,22 @@ void DataFlash_Class::Log_Write_Compass_instance(const Compass &compass, const u
...
 // Write a Compass packet
-void DataFlash_Class::Log_Write_Compass(const Compass &compass, uint64_t time_us)
+void DataFlash_Class::Log_Write_Compass(uint64_t time_us)
 {
     if (time_us == 0) {
         time_us = AP_HAL::micros64();
     }
-    Log_Write_Compass_instance(compass, time_us, 0, LOG_COMPASS_MSG);
+    const Compass &compass = AP::compass();
+    if (compass.get_count() > 0) {
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/DataFlash/LogFile.cpp]: I considered it :-)

Too many lines-changed for too-little advantage.

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp]: OK

File: APMrover2/APMrover2.cpp (hunk 3 above)
Comment (peterbarker) [reply APMrover2/APMrover2.cpp]: set_compass used to do `set_orientation` - which is renamed to `update_orientation`.  So I don't think this is new behaviour.

File: APMrover2/sensors.cpp (hunk 4 above)
Comment (peterbarker) [reply APMrover2/sensors.cpp]: https://github.com/ArduPilot/ardupilot/pull/8733/files#diff-65e09ab0b9a6008289271c4143ad740eL147

**Pull request closed without merging**